src/parser.cpp

```diff
@@ -106,6 +120,273 @@ Buffer Usage
 
+namespace {
+
+struct discontiguous_iterator
+{
+    buffers::const_buffer const* pos = nullptr;
+    std::size_t off = 0;
+
+    discontiguous_iterator(
+        buffers::const_buffer const* pos_)
+    : pos(pos_)
+    {
+    }
+
+    char
+    operator*() const noexcept
+    {
+        BOOST_ASSERT(pos);
+        auto it =
+            reinterpret_cast<char const*>(pos->data()) +
+            off;
+        return *it;
+    }
+
+    discontiguous_iterator&
+    operator++() noexcept
+    {
+        if( off >= pos->size() )
+        {
+            ++pos;
+            off = 0;
+            return *this;
+        }
+        ++off;
+        return *this;
+    }
+
+    discontiguous_iterator
+    operator++(int) noexcept
+    {
+        auto old = *this;
+        ++*this;
+        return old;
+    }
+
+    bool
+    operator==(
+        discontiguous_iterator const& rhs) const noexcept
+    {
+        return pos == rhs.pos;
+    }
+
+    bool
+    operator!=(
+        discontiguous_iterator const& rhs) const noexcept
+    {
+        return !(*this == rhs);
+    }
+};
+
+constexpr static
+std::size_t const max_chunk_header_len = 16 + 2;
+
+constexpr static
+std::size_t const last_chunk_len = 5;
+
+void
+parse_chunk_header(
+    buffers::circular_buffer& input,
+    system::error_code& ec,
+    std::size_t& chunk_remain_)
+{
+    if( input.size() == 0 )
+    {
+        ec = error::need_data;
+        return;
+    }
+
+    char tmp[max_chunk_header_len] = {};
+
+    core::string_view sv(
+        tmp,
+        buffers::buffer_copy(
+            buffers::mutable_buffer(
+                tmp, max_chunk_header_len),
+            input.data()));
+
+    auto it = sv.begin();
+    auto rv =
+        grammar::parse(it, sv.end(), detail::hex_rule);
+
+    if( rv.has_error() )
+    {
+        ec = error::bad_payload;
+        return;
+    }
+
+    auto rv2 =
+        grammar::parse(it, sv.end(), detail::crlf_rule);
+    if( rv2.has_error() )
+    {
+        if( rv2.error() == condition::need_more_input )
+            ec = error::need_data;
+        else
+            ec = error::bad_payload;
+        return;
+    }
+
+    if( rv->v == 0 )
+    {
+        ec = error::bad_payload;
+        return;
+    }
+
+    input.consume(it - sv.begin());
+    chunk_remain_ = rv->v;
+};
+
+void
+parse_last_chunk(
+    buffers::circular_buffer& input,
+    system::error_code& ec)
+{
+    // chunked-body = *chunk last-chunk trailer-section CRLF
+    // last-chunk = 1*"0" [ chunk-ext ] CRLF
+    //
+    // drop support trailers/chunk-ext, use internal definition
+    //
+    // last-chunk = 1*"0" CRLF CRLF
+
+    if( buffers::buffer_size(input.data()) <
+        last_chunk_len )
+    {
+        ec = error::need_data;
+        return;
+    }
+
+    auto cbs = input.data();
+    discontiguous_iterator pos = cbs.begin();
+    discontiguous_iterator end = cbs.end();
+
+    std::size_t len = 0;
+    for( ; pos != end && *pos == '0'; ++pos, ++len )
+    {
+    }
+
+    if( len == 0 )
```

Comment:
I'm confused... len starts out 0, and the loop terminates if the first digit is not '0'. So this will then return "bad payload?"

Reply:
When we call this function, it's because we're on a chunk boundary. We're either looking for a chunk-size _or_ we're looking for a last-chunk. This function is called second, after we probe for a chunk-size.

The function also early exits with `needs_data` if we don't have a minimum of 5 characters, so we know at this stage of the code that we have data and none of it is valid.

To me, this seemed to imply an invalid payload but we can change this all up.